src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+        let (start_idx, end_idx) = (ffi_src.find('{').unwrap(), ffi_src.rfind('}').unwrap());
+        &ffi_src[start_idx+1..end_idx]
+    };
+    // Remove all whitespace and split `;` to get an Iterator over the decls:
+    let input: Vec<String> = input.split(';').map(ToString::to_string).collect();
+
+    // Parse each declaration and generate wrappers
+    let mut decls = Vec::new();
+    for decl in &input {
+        if decl.trim().is_empty() { continue; }
+        let (doc, attrs) = {
+            let mut doc = Vec::new();
+            let mut attrs = Vec::new();
+            for line in decl.lines() {
+                let line = line.trim();
+                if line.starts_with("///") {
+                    doc.push(line.to_string());
+                } else if line.starts_with("#") {
+                    attrs.push(line.to_string());
+                } else {
+                    break;
+                }
+            }
+            (doc, attrs)
+        };
+        let decl = decl.replace(|c| c == '\n' || c == '\r', "");
+        let is_pub = decl.find("pub fn").is_some();
+        let start_idx = decl.find("fn ").unwrap();
+        let decl = &decl[start_idx..];
+        let name = {
+                let end_idx0 = decl.find('<');
+                let end_idx1 = decl.find('(');
+                let end_idx = match (end_idx0, end_idx1) {
+                    (Some(v), None) => v,
+                    (None, Some(v)) => v,
+                    (Some(i), Some(j)) => i.min(j),
+                    (None, None) => panic!(),
+                };
+                decl["fn ".len()..end_idx].to_string()
+        };
+        match name.as_str() {
+            | "jpeg_float_add_quant_table"
+                | "jpeg_calc_jpeg_dimensions"
+                | "jsimd_fdct_ifast"
+                | "jsimd_can_rgb_ycc"
+                | "jsimd_can_fdct_ifas"
+                => continue,
+
+            _ => (),
+        }
+        let lifetime = decl.find('<').map(|v| decl[v+1..decl.find('>').unwrap()].to_string());
+        let args: Vec<(String,String)> = decl.find('(')
+                .map(|v| &decl[v+1..decl.find(')').unwrap()]).unwrap().split(',')
+            .filter(|s| !s.trim().is_empty())
+            .map(|s|{
+                let mut s = s.split(':');
+                let arg_name = s.next().unwrap();
+                let ty_name = s.next().unwrap();
+                (arg_name.trim().to_string(), ty_name.trim().to_string())
+            }).collect();
+        let ret_ty = decl.rfind("->").map(|s| decl[s+2..].trim().to_string());
+        decls.push( Decl { name, is_pub, doc, attrs, args, ret_ty, lifetime });
+    }
+
+    decls
+}
+
+fn generate_cxx_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_cxx_wrapper(d)).collect()
+}
+
+fn generate_rust_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_rust_wrapper(d)).collect()
+}
+
+fn generate_rust_wrapper(decl: &Decl) -> String {
```

Comment:
This is half of the relevant code. If you use syn for parsing, you can just use the `quote`crate to make this quite a bit nicer.

In a nutshell, instead of exposing `extern "C"` decls directly from `mozjpeg-sys`, what you expose is thin functions that call the declaration, and transform whatever error you get from C, in whatever error you want in Rust.